### PR TITLE
fix: use correct course_key in case of a key_for_rerun in RCM

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-4133](https://2u-internal.atlassian.net/browse/PROD-4133)

Updates the logic behind bringing the courses from studio in case a course is created in Studio and not in Discovery. Previous logic used to create a course_run from Studio into Discovery with whatever was present in the key. In case a course has a value in `key_for_reruns`, it used to assign those course_runs to the key equal to `key_for_reruns`.

Now, it finds if a course has a `key_for_reruns`, it will assign this course_run to the correct course that has the custom `key_for_reruns` value.

### CODE REVIEW

File: course_discovery/apps/course_metadata/data_loaders/api.py [hunk 1]
```diff
@@ -202,6 +202,8 @@ def create_course_run(self, course, body):
     def get_or_create_course(self, body):
         course_run_key = CourseKey.from_string(body['id'])
         course_key = self.get_course_key_from_course_run_key(course_run_key)
+        course_key = (Course.objects.filter(key_for_reruns=course_key).values_list('key', flat=True).first() or
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/api.py]: nit: There should be a comment explaining this scenario in the code. Or better, have an internal or util function to return the course key that considers both scenarios.

File: course_discovery/apps/course_metadata/data_loaders/tests/test_api.py [hunk 2]
```diff
@@ -534,6 +534,32 @@ def test_ingest_studio_made_run_with_existing_draft_course(self):
...
+    @responses.activate
+    def test_ingest_with_key_for_rerun(self):
+        """
+        Verify that we correctly stitch up a course run made in Studio but having key_for_rerun.
+        The test makes sure that the course_run isn't incorrectly assigned to course having key = key_for_rerun.
+        """
+        datum = mock_data.COURSES_API_BODY_ORIGINAL
+        self.mock_api([datum])
+        course_type = CourseType.objects.get(slug=CourseType.AUDIT)
+        course_key = '{org}+{number}'.format(org=datum['org'], number=datum['number'])
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/test_api.py]: you can use f-string, no need to use format anymore.

File: course_discovery/apps/course_metadata/data_loaders/tests/test_api.py [hunk 3]
```diff
@@ -534,6 +534,32 @@ def test_ingest_studio_made_run_with_existing_draft_course(self):
...
+        course_key_with_key_for_rerun = 'test+key_for_rerun'
+        Course.objects.create(partner=self.partner, key=course_key, title='Title', type=course_type)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/test_api.py:548]: should the first course be created here? Shouldn't only the course with key_for_reruns be created only?

Comment (Ali-D-Akbar) [reply course_discovery/apps/course_metadata/data_loaders/tests/test_api.py:548]: The scenario we're testing should have a courseA and courseB having key_for_reruns = courseA.key

**Review by DawoudSheraz — APPROVED** [review]

**Review by AfaqShuaib09 — APPROVED** [review]

**Pull request merged**